runtime/context.go

```diff
@@ -61,6 +61,12 @@ func AnnotateContext(ctx context.Context, req *http.Request) (context.Context, e
 			if strings.ToLower(key) == "authorization" {
 				pairs = append(pairs, "authorization", val)
 			}
+			if mux.headerMatcher != nil {
+				if h, ok := mux.headerMatcher(key); ok {
+					pairs = append(pairs, h, val)
+					continue
+				}
+			}
 			if isPermanentHTTPHeader(key) {
```

Comment:
It seems to me that isPermanentHTTPHeader and the following MetadataHeaderPrefix check are just the default forms of the header matcher. What would you think about moving them into a HeaderMatcher and providing them an overridable default headermatcher? That way we can unblock a lot of the bugs around headers not behaving the way people expect?